### ISSUE TITLE
tf.set_random_seed does not reset random op state

### ISSUE THREAD

**Issue by msmsajjadi** [tensorflow/tensorflow#9171]
TF Version: 1.1.0rc1 (installed from nightly: `Apr 10, 2017 1:03 AM`)
(run on CPU, Python 2)

```
import tensorflow as tf
import numpy as np

sess = tf.InteractiveSession()
sess.run(tf.global_variables_initializer())

tf.set_random_seed(1)
a = tf.truncated_normal_initializer(seed=None)([1])
print(a.eval())

tf.set_random_seed(1)
b = tf.truncated_normal_initializer(seed=None)([1])
print(b.eval())
```

Output:
```
[ 1.05293429]
[-0.4487586]
```

Expected:
The same value, since...
> If the graph-level seed is set, but the operation seed is not: The system deterministically picks an operation seed in conjunction with the graph-level seed so that it gets a unique random sequence.

The values are identical in repeated runs of the whole script, but not after resetting the graph-level seed (as in the example above).

(possibly related to https://github.com/tensorflow/tensorflow/issues/9003)

**Comment by jart** [comment]
The [release notes](https://github.com/tensorflow/tensorflow/blob/master/RELEASE.md) say determinism is only promised for `tf.set_random_seed(0)`. Does using `0` fix things for you? If not, then please tell me more and I'll re-open.

**Comment by seaotterman** [comment]
Thanks, unfortunately it doesn't seem to work with seed=0 either, tested on the latest nightly from `Build #457 (Apr 14, 2017 1:03:00 AM)` (nightly whl binary downloaded from the repository, for Mac, CPU-only version, Python 2.)

I get the following output for the above code when using seed 0 in both cases:
```
[ 1.53692079]
[ 1.01875198]
```

Is there a reason why it should only be deterministic for that specific seed?

**Comment by seaotterman** [comment]
@jart I found the relevant part of the code and I'm suspecting this to be intended behavior. I'll open a new ticket concerning the random seed logic.

**Comment by jart** [comment]
Or we can reopen this one, as promised. Let me find someone to rope into this conversation while you do some extra investigation.

**Comment by jart** [comment]
@girving Is the code example in this bug relating to `tf.set_random_seed(0)` nondeterminism WAI? See: https://github.com/tensorflow/tensorflow/commit/e9786df5e89f0345b2eb32d688c7be31c5259ba0

**Comment by seaotterman** [comment]
Alright!

Simpler, more low-level code example:
```
for i in range(2):
    tf.set_random_seed(1)
    print(tf.random_uniform([1], seed=None).eval())
```
Output:
```
[ 0.77878559]
[ 0.54316521]
```
Judging from on the current implementation, this seems to be intended behavior. In `random_seed.py`, `op_seed` is set to `ops.get_default_graph()._last_id` if the passed parameter for `op_seed` is `None`. This value depends on the execution and can change (as it does in the example above).

The wording in the documentation could be read either way:
  3. If the graph-level seed is not set, but the operation seed is set:
    A default graph-level seed and the specified operation seed are used to
    determine the random sequence.

I think the following behavior would be more natural:
- `graph_seed` is changed randomly upon each creation of a random variable (current implementation: `graph_seed` is fixed!)
- if `op_seed` is `None`, use only `graph_seed` for the creation of the random values
- if `op_seed` is given, use only `op_seed` (and ignore `graph_seed`)

Motivation:
There doesn't seem to be an elegant way to have, say, 2 neural network models
initialized to the same random values in the same execution graph. Current
workarounds include the usage of `update_ops` to copy values from one network to
the other, or to set the `op_seeds` for all operations which leads to
hard-to-maintain code (because the `op_seeds` would need to be deterministic but
different for each layer to ensure that we don't get the same values for all
variables).

**Comment by girving** [comment]
Correct, a seed of `None` without a graph level seed is intentional nondeterministic behavior.  Have you considered setting the seed to a value other than `None`?  If you set two random ops to the same seed, they will produce the same stream of random numbers.

**Comment by seaotterman** [comment]
@girving 
> Correct, a seed of None without a graph level seed is intentional nondeterministic behavior.
But the graph level seed is set in the example. Fixing the op seeds is possible, but then the user needs to produce pseudo-random op_seeds for all parts of the code to avoid identical random values in the same section (see also the motivation part of my reply above).

In other words, the graph seed is currently only useful for identical results over repetitions of the whole code, while the proposed solution would be more flexible and make the creation of identical values much easier.

**Comment by girving** [comment]
Ah, your issue is poorly named; fixed.  The thing you want is still impossible, since the state you're trying to set is not part of the graph: it is part of the session.  Resetting the state on the session would be a nice feature (@langmore wants it too).

The other option is to use random ops with custom seed control, which I am about to add as `tf.contrib.stateless.stateless_random_uniform`, etc.

**Comment by girving** [comment]
To set the state on the session and reuse the existing random ops, we would need some plumbing to query the OpKernels for their internal state.

**Comment by girving** [comment]
Oh, I see: yes, the thing you want to do is possible after all; we'd just have to make the default per-op seeds count from the last time `tf.set_random_seed` is called.

**Comment by seaotterman** [comment]
That would be great!

**Comment by girving** [comment]
Well, I didn't volunteer to do it. :)

@michaelisard, @aselle: Do you have thoughts about this?  Modifying `tf.set_random_seed` to reset the per-op seed counter is trivial to do, but it would probably be quite surprising to users when it didn't work to reset states after graph construction is complete.

**Comment by seaotterman** [comment]
What about a dedicated function `reset_op_seeds` which can only be called during graph construction? Or is it possible to reset those seeds already (manually)?

**Comment by seaotterman** [comment]
It appears @girving's stateless random ops are available in TF 1.2rc0:
https://www.tensorflow.org/versions/master/api_docs/python/tf/contrib/stateless

This is a useful addition, however it does not simplify the use case of initializing 2 neural network models with identical weights since the weights are generally built via tf.layers/tf.contrib modules which call stateful random-ops. I might be missing an obvious easy solution for this, but I guess stackoverflow is the place for those questions.

Any updates @michaelisard @aselle?

**Comment by michaelisard** [comment]
@asimshankar assigning to you for triage. It seems as if the current way we set random seeds has missing use cases: do you want to look at future redesigns?

**Comment by asimshankar** [comment]
I think adding a `reset_op_seed` parameter to `tf.set_random_seed` could be acceptable. Something like: `tf.set_random_seed(seed, reset_op_seed=False)` (I'll ask around if anyone has objections for it defaulting to `True`).

@ekelsen has been looking at determinism in general and might have other ideas.

@girving : RE [your comment](https://github.com/tensorflow/tensorflow/issues/9171#issuecomment-294188068) - I don't think it would be any more or less surprising for the op-seed reset to not take effect after graph construction is complete than it is for `tf.set_random_seed` to have no effect after graph construction is complete today. Or am I missing something?

**Comment by hillst** [comment]
any progress on this? looking for a simple approach to implement dropout as a bayesian representation, https://arxiv.org/pdf/1506.02142.pdf

**Comment by tensorflowbutler** [comment]
It has been 14 days with no activity and this issue has an assignee.Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assigneee: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by frthjf** [comment]
I ran into the same issue. However, for my use case I was able to use the stateless randomness:

```python
for i in range(2):
    seed=(1,1)
    print(tf.contrib.stateless.stateless_random_uniform([1], seed).eval())
```

**Comment by seaotterman** [comment]
@frthjf These new additions are useful, but as mentioned above, they are not being used by the common layer building blocks (yet?), so one would need to pass these manually to all layers (including special initialization schemes).

**Comment by tensorflowbutler** [comment]
Nagging Assignee @asimshankar: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @asimshankar: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by TYS11** [comment]
@seaotterman have you found a way to initialize the same weights for Neural Network multiple times over different session?

**Comment by tensorflowbutler** [comment]
Nagging Assignee @asimshankar: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @asimshankar: It has been 29 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @asimshankar: It has been 44 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @asimshankar: It has been 61 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @asimshankar: It has been 76 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @asimshankar: It has been 91 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @asimshankar: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @asimshankar: It has been 29 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @asimshankar: It has been 44 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @asimshankar: It has been 59 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @asimshankar: It has been 74 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @asimshankar: It has been 89 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @asimshankar: It has been 104 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @asimshankar: It has been 122 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @asimshankar: It has been 136 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by wangpengmit** [comment]
We are addressing this problem by a comprehensive revamp in RFC tensorflow/community#38. Please comment on that.

**Comment by crobarcro** [comment]
> We are addressing this problem by a comprehensive revamp in RFC [tensorflow/community#38](https://github.com/tensorflow/community/pull/38). Please comment on that.

Just to be clear, does this mean nothing will happen on this issue until v2.0 is released? It really hinders debugging.

**Comment by wangpengmit** [comment]
> > We are addressing this problem by a comprehensive revamp in RFC [tensorflow/community#38](https://github.com/tensorflow/community/pull/38). Please comment on that.
> 
> Just to be clear, does this mean nothing will happen on this issue until v2.0 is released? It really hinders debugging.

I'm afraid so.

**Comment by lostmsu** [comment]
What are my options prior to 2.0? Do I need to create a new graph every time I need to reproduce a result with some fixed seed?

**Comment by wangpengmit** [comment]
> What are my options prior to 2.0? Do I need to create a new graph every time I need to reproduce a result with some fixed seed?

Always providing the 'seed' argument to ops that accept it, along with manually setting the global seed via `tf.set_random_seed`, can mitigate the problem to some degree.

**Comment by lostmsu** [comment]
Is there a way to reset all the internal state in the ops to its original value without recreating graph?

**Comment by wangpengmit** [comment]
> Is there a way to reset all the internal state in the ops to its original value without recreating graph?

No, the op's state is stored in a C++ member variable with no access from outside. Exposing it is among the main motivations of the RFC.

**Comment by wangpengmit** [comment]
I'm closing this since the new RNGs (https://github.com/tensorflow/tensorflow/blob/master/tensorflow/python/ops/stateful_random_ops.py) are ready.

**Comment by AarynBaelish** [comment]
try to use `tf.reset_default_graph()`

```python
import tensorflow as tf  
for i in range(2):  
    with tf.Session() as sess:  
        tf.set_random_seed(2)  
        var = tf.Variable(tf.random_normal([1, 1], 0.0, 0.01))  
        init = tf.global_variables_initializer()  
        sess.run(init)  
        print(var.eval())  
    tf.reset_default_graph() 
```

Output:
```
[[-0.0142362]]
[[-0.0142362]]
```
